bpf/bpf_lxc.c

```diff
@@ -902,7 +902,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 			if (eth_store_daddr(ctx, (__u8 *)&vtep->vtep_mac, 0) < 0)
 				return DROP_WRITE_ERROR;
 			return __encap_and_redirect_with_nodeid(ctx, vtep->tunnel_endpoint,
-								WORLD_ID, &trace);
+								SECLABEL, WORLD_ID, &trace);
```

Comment:
For these cases where both SECLABEL (`selabel`) and WORLD_ID (`vni`) are passed down, this change makes sense to me. I'm not sure I understand all of the other cases where the vni is passed down as `0`. What's the intention there, why don't those paths need to override the VNI? Is it because they're only for traffic being routed to other nodes in the cluster?

If that's the case, then we might be able to better-document this by doing a 
```
/* NO_VTEP_DST is passed to an encapsulation function when the destination of the tunnel is not a VTEP. */
#define NO_VTEP_DST 0
```

Then pass that from all of the functions that are passing 0 today (assuming this holds true, those paths don't hand the traffic to the VTEP), and update the `if (vni)` statement to `if (vni != VTEP_NO_DST)`.

What do you think?

Reply:
correct me if I am wrong,  I should put `#define NO_VTEP_DST 0` in `bpf/node_config.h`

Reply:
I think it's fine to put it into `bpf/lib/encap.h`. `node_config.h` is generated at runtime, so if you put it there then you'll have to update the agent code as well. But we won't be dynamically changing that value.

Reply:
pushed the new changes and tested, client2 pod with L7 egress policy allowed to VTEP

egress

```
Ethernet        {Contents=[..14..] Payload=[..70..] SrcMAC=fa:2d:34:7f:91:c8 DstMAC=82:36:4c:98:2e:56 EthernetType=IPv4 Length=0}
IPv4    {Contents=[..20..] Payload=[..44..] Version=4 IHL=5 TOS=0 Length=64 Id=61216 Flags=DF FragOffset=0 TTL=64 Protocol=TCP Checksum=13542 SrcIP=10.0.1.171 DstIP=10.1.1.6 Options=[] Padding=[]}
TCP     {Contents=[..44..] Payload=[] SrcPort=53058 DstPort=8080(http-alt) Seq=1173472904 Ack=2058451550 DataOffset=11 FIN=false SYN=false RST=false PSH=false ACK=true URG=false ECE=false CWR=false NS=false Window=502 Checksum=5860 Urgent=0 Options=[..6..] Padding=[]}
CPU 01: MARK 0xd225c90a FROM 3202 to-overlay: 78 bytes (78 captured), state unknown, interface cilium_vxlan, , identity 12073->unknown, orig-ip 0.0.0.0

```
ingress

```
Ethernet        {Contents=[..14..] Payload=[..62..] SrcMAC=0e:61:5b:57:3c:24 DstMAC=2a:a2:48:49:75:ac EthernetType=IPv4 Length=0}
IPv4    {Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=0 Flags=DF FragOffset=0 TTL=63 Protocol=TCP Checksum=9458 SrcIP=10.1.1.6 DstIP=10.0.1.196 Options=[] Padding=[]}
TCP     {Contents=[..40..] Payload=[] SrcPort=8080(http-alt) DstPort=58468 Seq=3631714514 Ack=363024372 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=true URG=false ECE=false CWR=false NS=false Window=65160 Checksum=5881 Urgent=0 Options=[..5..] Padding=[]}
CPU 03: MARK 0x66156b7a FROM 3671 to-endpoint: 74 bytes (74 captured), state reply, interface lxc490714591e9b, , identity world->12073, orig-ip 10.1.1.6, to endpoint 3671


CPU 05: MARK 0xf4985f2 FROM 3671 DEBUG: Conntrack lookup 1/2: src=10.0.1.196:58468 dst=10.1.1.6:8080
-> Response http to 3671 ([k8s:io.kubernetes.pod.namespace=cilium-test k8s:kind=client k8s:name=client2 k8s:other=client k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=cilium-test k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default]) from 0 ([reserved:world]), identity 2->12073, verdict Forwarded GET http://10.1.1.6:8080/ => 200

```

Reply:
For egress, I see `identity 12073->unknown`. I would have expected that to say something like `identity 12703->2` ie sending to world. Do you know why it shows up like that?

Reply:
ah, we probably should set  `dst` label to `WORLD_ID` in `send_trace_notify` if `ENABLE_VTEP` enabled

```
#ifdef ENABLE_VTEP
        if (vni != NOT_VTEP_DST)
                key.tunnel_id = vni;
        else
#endif /* ENABLE_VTEP */
                key.tunnel_id = seclabel;
        key.remote_ipv4 = node_id;
        key.tunnel_ttl = 64; 
                  
        cilium_dbg(ctx, DBG_ENCAP, node_id, seclabel);
        
        ret = ctx_set_tunnel_key(ctx, &key, sizeof(key), BPF_F_ZERO_CSUM_TX);
        if (unlikely(ret < 0))
                return DROP_WRITE_ERROR;
                                 
        send_trace_notify(ctx, TRACE_TO_OVERLAY, seclabel, 0, 0, ENCAP_IFINDEX,
                          ct_reason, monitor);
        return 0;
```

Reply:
`send_trace_notify` in `__encap_with_nodeid` has always been setting dst secure label to `0/unknown`, in this PR I could set dst secure label to `WORLD_ID` if `ENABLE_VTEP` enabled, leave it to `0` for now, later we could maybe refactor the encap function to pass on dst secure label from encap function callers if we think this is a bug @joestringer

Reply:
Good find. Yeah we can follow up. Also I'm not sure that just setting `WORLD_ID` is a good idea for this, there may be a more specific locally allocated CIDR identity for the destination, which the ipcache would look up. Ideally we'd pass that identity along to the encap function here so that we can report the correct destination identity.

But yeah I'm fine to defer this for now.

Reply:
tried 
```
__u32 dst_seclabel = 0;
#ifdef ENABLE_VTEP
        if (vni != NOT_VTEP_DST) {
                key.tunnel_id = vni;
                dst_seclabel = WORLD_ID;
        } else
#endif /* ENABLE_VTEP */
              key.tunnel_id = seclabel;
```
checkpatch complains unbalanced `{`, any idea to work around this?

Reply:
Maybe wrap the last line here in an indented section?

```
{
        key.tunnel_id = seclabel;
}
```

Reply:
```
#ifdef ENABLE_VTEP
        if (vni != NOT_VTEP_DST) {
                key.tunnel_id = vni;
                dst_seclabel = WORLD_ID;
        } else
#endif /* ENABLE_VTEP */
        {
                key.tunnel_id = seclabel;
        }
```

complains

```
CHECK:BRACES: Unbalanced braces around else statement
#109: FILE: bpf/lib/encap.h:141:
+	} else
 
```
I had tried using  two `#ifdef ENABLE_VTEP` for  `key.tunnel_id` and  `dst_seclabel` separately, but just seems redundant

Reply:
> Good find. Yeah we can follow up. Also I'm not sure that just setting `WORLD_ID` is a good idea for this, there may be a more specific locally allocated CIDR identity for the destination, which the ipcache would look up. Ideally we'd pass that identity along to the encap function here so that we can report the correct destination identity.
> 
> But yeah I'm fine to defer this for now.

yes, the `*dst_id` is already available for example in bpf_lxc

```
        /* Determine the destination category for policy fallback. */
        if (1) {
                struct remote_endpoint_info *info;

                info = lookup_ip4_remote_endpoint(ip4->daddr);
                if (info && info->sec_label) {
                        *dst_id = info->sec_label;
```

Reply:
https://github.com/cilium/cilium/issues/20169 to track this issue

Reply:
One thought I didn't look into, what do the callers pass down if VNI support is disabled? ie when they pass `vni` but `!defined(ENABLE_VTEP)`, do they just pass down `0` / `NOT_VTEP_DST`? If that's the case, then this function might not need a macro at all.

Reply:
^ But no need to look into above idea in this PR, we can follow up ons that as part of #20169.